### DISCUSSION TITLE
Http4s site discussion

### DISCUSSION THREAD

**Discussion by keynmol** [http4s/http4s#5192] [General]
Continuing a discussion started by Noel Welsh on twitter, hoping to discuss some ideas here and break them out into issues once we reach consensus.

What follows is a living list of concrete suggestions to improve the documentation, both aesthetics and content

## Stylistic

Versions page:
- Only keep 0.21, 0.22, 0.23, 1.x
  
  Put "Which version is right for me" next to lineage rows

- Remove Scala 2.10, 2.11 from the table

Whole site:
- Design
   
   https://github.com/http4s/http4s/issues/5191

- Add Algolia DocSearch: https://www.algolia.com/
  
  See https://github.com/sbt/website/pull/963 for whole process, should be simpler with Hugo (and no multilingual sites)

- Use `mdoc:compile-only` for most places, to avoid the rendering of `IO` values.

  Consider doing something like mdoc-effect: https://mdoc-effect.indoorvivants.com/ce3/index.html to separate output from the code

- (minor) formatting of code examples

   Pretty snippets are pretty. Scalafmt 3.0.0 introduces that support for scala-fenced snippets: https://github.com/scalameta/scalafmt/releases/tag/v3.0.0

- Explicit links to sections
  
  Perhaps I'm dense, but hopefully this can be done with Hugo (i.e. like Cats effect site does it: 
<img width="96" alt="image" src="https://user-images.githubusercontent.com/1052965/133296945-fc26145d-a9bd-4541-8fcb-d6b504456e8f.png">) 
  
  This should help deep linking

# Content

- Explanation what different imports do 
  
   It's probably a hot topic in future versions of http4s, but for now it'd be good to provide a troubleshooting guide for common errors that are fixed by `dsl.io._` or `implicits` imports, along with explanations. Hopefully that can be indexed by search engines

- Update Quick start

   My feeling is that the screen space on Quick start page is limited and very expensive, so perhaps the listing of giter8 variables can be moved somewhere further down, and instead an example of Http4s setup in a single-file `IOApp` with server and client would be better?

**Comment by noelwelsh** [comment]
#5191 for simple design changes.

I would also like to:

* Invert the pyramid so the most important things come first. For example, the JSON page currently discusses something that no-one uses (raw JSON) before describing the things that people will actually use. It should be the other way around.

**Comment by keynmol** [comment]
There's also a fair number of issues labeled with `docs` that cover a lot of previous user requests: https://github.com/http4s/http4s/issues?q=is%3Aopen+is%3Aissue+label%3Adocs

**Comment by noelwelsh** [comment]
I'm going to look at updating Hugo to the latest version. 

(For the record, I will resent everything I have to learn about Nix.)

Update: I'm going to peace out on learning nix because it's just making me angry. If someone can change the hugo dependency to 0.88.1 and get nix-shell to actually recognize that change I'll rebuild the site from there.

**Reply by rossabaker** [reply]
I can get the nix-shell on hugo-0.88.1.  Are you still of the opinion that's a good tool, or did you prefer something different?  Two JVM-only solutions that mean less faffing about with Nix include:

* Paradox, used by many Lightbend projects and our jdk-http-client.
* Laika, whose author expressed an interest in helping Typelevel projects

There's also sbt-microsites, but if you think Nix makes you mad, you should see how mad jekyll makes me.

**Reply by armanbilge** [reply]
@tpolecat uses paradox too in his projects IIRC.

**Reply by noelwelsh** [reply]
I don't have a strong opinion here. I have used many static site generators in the past (Jekyll, Hakyll, Zola, Eleventy, Next, Gatsby) and they all annoy me in various ways. I have used Paradox for the [Doodle documentation](https://www.creativescala.org/doodle/). It's quite rigid in the organisational structure it allows but it has some nice features like linking to scaladoc and I think it's good enough for http4s.  I haven't used [Laika](https://planet42.github.io/Laika/0.18/table-of-content.html). It looks a bit more flexible.

What are the requirements you'd like? I guess we're aiming for low cognitive overhead / a minimum of things that can break?

I can spike out Laika integration is that is helpful.

**Reply by rossabaker** [reply]
Desiderata:
1. Really static generators. Things where we don't have to contemplate how accessible the JavaScript rendering of an SPA is.
2. Amenable to search. @keynmol has mentioned Algolia. If the search requires scripting, that's probably fine.
3. Something that I personally don't have to theme.  Off the shelf or volunteer contribution are fine.  I suck.
4. Few moving parts.  A tiny minority of our contributors embrace nix shell, a parallel setup is necessary on CI unless we thrust nix upon all maintainers, and everybody else has to install global plugins that may clash with their other projects.  JVM is best, static binaries like Hugo and Zola are second best, Jekyll and the JavaScript monstrosities need a strong argument.
5. Fits within the GitHub Pages feature set.  I'd like to move off Netlify, because they don't have a good community offering.  We'll lose rewrites.

**Reply by noelwelsh** [reply]
Ok, I will look at Laika. I'm assuming the documentation build can be automated with Github Actions.

My goal will be to reconstruct the current site in Laika but I will not attempt to reconstruct the current theme unless you are strongly attached to it.

**Reply by rossabaker** [reply]
I see no reason the doc build can't be automated with Actions.  I am not the slightest bit attached to the theme.

**Reply by armanbilge** [reply]
Just to throw one more idea into the ring: perhaps use the new static site generator in Scaladoc?
https://docs.scala-lang.org/scala3/scaladoc.html

Apparently it's capable of generating the entire Cats microsite. See:
Example: https://scala3docstorage.blob.core.windows.net/cats/index.html
Discussion: https://discord.com/channels/632277896739946517/632277897448652844/890943742054367253

I think this would go a long ways to meet Ross's few-moving-parts criterion.

**Reply by rossabaker** [reply]
Last I checked this feature had been added and removed, but that was in the pre-release cycle.  That's certainly a contender worth contemplating if it's real now.

**Reply by armanbilge** [reply]
Pinging @pikinier20, they made the example I posted above.

**Reply by noelwelsh** [reply]
Laika update:

* The sbt plugin for 0.18 conflicts with dependencies for other plugins (dependencies for Cats Effect, http4s)
* 0.17.1 works
* There is a parsing bug in code blocks that prevents further progress: https://github.com/planet42/Laika/issues/219

**Reply by pikinier20** [reply]
Hello, to be honest, there's a small misunderstanding about Cats site generation. I used sbt-microsite to do that. I only wanted to show that Scaladoc generates API for Cats with only some small changes in sbt and also that it works with sbt-microsite. However, after reading the discussion I think that you can maybe be interested in the preprocessor for Markdown files that @BarkingBad is currently working on. It i.e. adds links to Scaladoc, checks snippets. Also, it can contain other features since it's still in dev. After preprocessing you can use any static site engine while preserving features added by preprocessor.

**Reply by armanbilge** [reply]
Ah, I was confused. Thanks for chiming in and clarifying.

**Reply by BarkingBad** [reply]
The PR related to the changes can be found here https://github.com/lampepfl/dotty/pull/13140
It's not worth reviewing yet since it depends on another PR and will probably be reworked, but the basic idea is to preprocess the markdowns in the context of scaladoc and return them to be used by another templating tool (like jekyll). However, there is an option to create the whole static-stie from the scaladoc (for example nightlies of scaladoc are created that way https://dotty.epfl.ch/docs/index.html)

**Reply by noelwelsh** [reply]
Thanks to @jenshalm's help I've converted the documentation to Laika. The commit message on this branch explains what to do to preview, if you're interested: https://github.com/noelwelsh/http4s/tree/feature/laika

**Reply by jenshalm** [reply]
Sorry, chiming in late, I have missed the previous discussion, Laika creator here...

I had a quick look at @noelwelsh's branch. It is a good starting point, but there would need to be some extra work for a really polished site. 
Again, I'm happy to help with this and continue the work Noel has started by doing any of the steps described below. 
I'll list everything that immediately sprang to mind, but there might be more once I dig deeper:

* Noel had to use the CE2 branch of Laika due to dependency conflicts. I noticed that the http4s build explicitly depends on http4s 0.22.x. 
I wonder whether this can be changed or has to remain for now. 
Unfortunately the CE2 branch of Laika is no longer maintained and does not have feature parity with the CE3 branch.

* Titles do not have to be configured explicitly in Laika in case you want to use the first Level-1 headline of a page for navigation.
Hugo seems to inject the title right into the page, but Laika does not do that. 
For Laika it would be better to simply add an h1 onto each page and then omit the configuration.

* One immediate win with using Laika is page navigation on the top right. I always missed that with the longer pages of the http4s site.
The right navigation box can also contain a link to the page source, you'd only need to configure the base URL for this.

* You are using variables in some places in the form of `{{< version "http4s.doc" >}}`. 
I was unable to figure out where the actual value is injected, but the syntax would need to be different for Laika.

* With the current setup of the demo branch you are losing syntax highlighting. There are two choices for bringing it back. 
One is to use Laika's own internal syntax parser. I would claim most people won't notice the difference, you can browse Laika's docs for an impression of how it looks.
If you don't want that you can also add back external JS-based highlighters by adjusting the default template.

* Laika supports arbritary nesting of directories which are automatically reflected in the nesting levels of the navigation.
You could, for example, provide all middleware docs in a separate section. 
I'd feel that would be easier to browse than a long, flat list of links, but is ultimately a matter of preference, of course.

* The top navigation bar is by default nearly empty. You might want to add back all the links you currently have 
(btw. you are still linking to gitter, not Discord).

* Laika supports versioned documentation out of the box, it would just require some additional configuration.
 It also has a few advantages over what you have now: 
 First, switching versions stays on the same page if it exists in both, source and target version.
Second, the version dropdown is populated by JSON, meaning older versions automatically see all newer versions. 
This does currently not seem to be the case, as some version are missing in some of the older docs.
(Btw. I had issues browsing some versions, which did not load the CSS for me).

* You can easily adjust the theme to your preferred choice of colors and fonts, 
e.g. to bring back the orange/gray color scheme of your original site if you care for that.
This does not require digging deep into CSS, but can be done with a few lines of config in the build.

* Laika supports providing downloads of the whole documentation as EPUB and PDF out of the box 
 (these features are also implemented in a JVM-only fashion, so no worries that it would bring back some external tools)
This also would only require a bit of additional configuration (and maybe the creation of a title image).

* Laika supports the generation of a table of content for the site, which includes some of the page's sections up to a configurable level.

**Reply by armanbilge** [reply]
That sounds amazing 😮 thanks for chiming in @jenshalm, and to @noelwelsh for kickstarting the conversion.

> I noticed that the http4s build explicitly depends on http4s 0.22.x.

This can almost definitely be updated. I think there is an issue about it somewhere. So it sounds like if we unblock that, you're happy to work with @noelwelsh on a PR doing the above? :)

**Reply by jenshalm** [reply]
@armanbilge Yes, getting rid of CE2 in your build would be really beneficial. You could use Laika 0.17 for now, but only the 0.18 series would give you a promise of reasonably quick patch releases in case of issues. 
Otherwise I can start with that work at any point, I'd only need some of the maintainers to be available for questions about your preferences (optional features, styling, etc.) and your current build setup (e.g. where the variables come from, how it's integrated with CI, published to GitHub, etc.).

**Reply by noelwelsh** [reply]
@jenshalm If you want to take over the Laika integration that is great. My work is only a proof of concept at this stage. In addition to all the points you listed above I left the Hugo integration in place and that would need to be stripped out. Note that the website project relies on Hugo as well and needs to be updated. If you're working on this I will switch to working on the content.

**Reply by jenshalm** [reply]
@noelwelsh yes, I am happy to take over. I'd also have to touch the Markdown sources to some degree, e.g. for variable injection, but the merge conflicts should hopefully be limited, unless you plan to move content around.

**Reply by armanbilge** [reply]
@jenshalm I unblocked the CE2/CE3 thing in https://github.com/http4s/http4s/pull/5300. The merge is still working its way up to the main branch, but while you're waiting you can just manually apply the patch.

**Reply by jenshalm** [reply]
@armanbilge thanks, I actually already started with all the things which work the same in 0.17. I'd suggest I'd do a WIP PR later today or tomorrow with all the obvious migration steps applied minus all the things where I still have questions and minus all optional features, so that discussion can move there?

**Reply by armanbilge** [reply]
Sounds good, looking forward to it!

**Reply by noelwelsh** [reply]
I'll wait a few days before I start on the content so shouldn't conflict with what you're doing.

**Reply by jenshalm** [reply]
@armanbilge I found something I wanted to check first before I continue, as it feels like a potential showstopper. I noticed that all your pages are hosted as an `index.html` in a separate folder, for prettier URLs I assume. Laika does not support this pattern as an auto-translation step during transformation and it's not something I could easily and quickly add. The only way to retain this structure with Laika at the moment would be to physically organize the Markdown sources in equivalent folders, too, or to delve fairly deep into Laika extension territory.

**Reply by armanbilge** [reply]
@jenshalm hmm, that's frustrating. It sounds like your idea to put the docs in separate folders would be okay though (unless this will cause other problems, e.g. with navigation?) so that at least we can preserve existing links ...

**Reply by jenshalm** [reply]
@armanbilge: Moving them into folders would certainly work and it's quick and easy to do. I'm just not sure it's ok with maintainers as it's not very convenient to work with. The extension option is there, too, but it would A) delay the PR for several days at least, and B) mean you'd have the extension plus a little test suite for it in your code base until Laika potentially supports it out of the box.

I wasn't aware people do this kind of thing and it has never been requested for Laika in 9 years, that's mainly why it's not supported, it's certainly doable otherwise.

**Reply by armanbilge** [reply]
@jenshalm I'm too new of a maintainer to give any authoritative answer, but it sounds like a little extra folder structure (although annoying) is much more tried-and-true than a custom extension in our docs project, which would be inconsistent with Ross's "few moving parts" criterion in https://github.com/http4s/http4s/discussions/5192#discussioncomment-1362022. If one day Laika could support this out of the box, that would be awesome, and we can revert the original structure. If it's too annoying, we could also just burn all of our existing links (I wonder if this could be mitigated with some quickie Javascript on the 404 page to redirect or even just a kind message apologizing for the restructuring).

**Reply by jenshalm** [reply]
@armanbilge Ok, thanks for chiming in, I'll add it to the list of open question on the PR then and wait for a definitive answer before moving on. There are other open questions, too, so I'm getting pretty close to needing feedback anyway.

**Comment by rossabaker** [comment]
@keynmol has many excellent ideas here, but let me pour some cold water on one:

> (minor) formatting of code examples

> Pretty snippets are pretty. Scalafmt 3.0.0 introduces that support for scala-fenced snippets: https://github.com/scalameta/scalafmt/releases/tag/v3.0.0

I've been closing all the scalafmt-3.0 PRs until we regain some branching sanity (which I'll be initiating another discussion on soon).  I tried running it on each branch, merging it with the previous branch, keeping ours on all conflicts, so we could proceed from a solid merge base.  It was a lot of work, and scalafmtCheck failed, so I reverted to 2.x.  It might be ready to try again now that it's a few patch releases deep.

**Reply by keynmol** [reply]
That's fair 👍 TBH branching is an issue for docs in a more general sense - as I understand it, adding docs for core concepts (that have changed in the 0.21 -> 0.22 -> 0.23 -> 1.0) would require 4 versions of docs for each branch to make sure they all compile?

**Reply by rossabaker** [reply]
Publishing compiled mdoc from each branch is what we've done in the past, but it complicates things with an ad hoc process.  I think it's nice to have, but one good branch of docs is better than four mediocre ones if we have to choose.

**Comment by kubukoz** [comment]
Made https://github.com/http4s/http4s/pull/5416 for some extra client docs. Working on this, I felt like there should be a table of contents for a given page: consider how many headings that site already has.

Also, copyable links for the headings would be a great thing to have. They already have `id`s so it's just a matter of adding a chain icon or just making them clickable ("copy link address" would work just fine).

**Reply by jenshalm** [reply]
@kubukoz See #5313, once this is on Laika, page navigation comes for free.

**Comment by noelwelsh** [comment]
Added #5539 for a tutorial
